### ISSUE TITLE
question: How do we know the classification output is accurate?

### ISSUE THREAD

**Issue by mainuddinMains** [oss-slu/baio#180]
## Question

When BAIO classifies a DNA sequence as **Virus** or **Host**, how do we verify that result is actually correct? This issue documents what accuracy signals exist today, what is missing, and what needs to be built so users and developers can trust the output.

---

## What the model currently reports per prediction

Every classification result already includes:

| Field | What it means |
|-------|--------------|
| `prediction` | `Virus` or `Host` |
| `confidence` | Probability score 0–1 (e.g. 0.87 = 87% confident) |
| `uncertain` | `true` if confidence is below the configured threshold (default 0.6) |
| `ood_score` | Out-of-distribution score — how different this sequence is from training data |

**Confidence alone does not equal accuracy.** A model can be 90% confident and still be wrong. The metrics below are what actually measure accuracy.

---

## Known accuracy numbers (current RandomForest + 6-mer model)

These are the only published numbers in the codebase (`docs/transformer_classification_guide.md`):

| Metric | Score | What it means |
|--------|-------|---------------|
| Cross-validation accuracy | 0.9575 | Accuracy on training folds — **optimistic** |
| **Test accuracy** | **0.71** | Accuracy on held-out data — **the real number** |
| ROC AUC | 0.9012 | Ability to rank Virus above Host |
| Virus recall | ~0.44 | Only 44% of actual viruses are correctly identified |

**Key concern:** The gap between CV accuracy (0.96) and test accuracy (0.71) indicates overfitting. The model misses more than half of real viral sequences.

---

## How to run the evaluation yourself

The evaluation framework already exists in `binary_classifiers/evaluation.py`. It computes:
- Accuracy, Precision, Recall, F1
- ROC AUC
- Confusion matrix
- Per-sequence error list (which sequences were wrong, with confidence scores)

To run it against your own labeled FASTA files:

```python
from binary_classifiers.evaluation import load_labeled_sequences, evaluate_predictor
from binary_classifiers.predict_class import PredictClass

predictor = PredictClass(model_name="RandomForest")
labeled = load_labeled_sequences(
    virus_file="path/to/virus.fasta",
    host_file="path/to/host.fasta"
)
report = evaluate_predictor(predictor, labeled)

print(f"Accuracy:   {report['accuracy']}")
print(f"Recall:     {report['recall']}")
print(f"F1:         {report['f1']}")
print(f"ROC AUC:    {report['roc_auc']}")
print(f"Errors:     {len(report['errors'])} misclassified sequences")
```

> **You need labeled FASTA files** — one file of known virus sequences and one of known host sequences. The model is tested against these ground-truth labels.

---

### 1. No `/evaluate` API endpoint
There is no way to submit labeled sequences through the UI or API and get back accuracy metrics. The evaluation code exists in Python but is not exposed to users.

### 2. No benchmark dataset committed to the repo
The `data/` folder exists but no standard held-out test set is committed. Without a fixed test set, results cannot be reproduced or compared across model versions.

### 3. No confidence calibration check
A confidence score of 0.87 should mean the model is right ~87% of the time. This has never been measured. The model may be systematically overconfident or underconfident.

### 4. No per-sequence explanation in the UI
The UI shows a confidence bar but does not show which k-mer features drove the prediction. Users cannot tell *why* a sequence was called Virus vs Host.

### 5. No regression tracking
There is no check that model accuracy stays the same (or improves) across code changes. A bad commit could silently degrade accuracy.

---

## Tasks

- [ ] **Commit a fixed test dataset** — a labeled set of virus + host FASTA sequences that serves as the permanent accuracy benchmark
- [ ] **Add `/evaluate` API endpoint** — accepts labeled sequences, returns accuracy/recall/F1/ROC AUC/confusion matrix
- [ ] **Add evaluation results to the UI** — show a summary accuracy panel when ground-truth labels are provided
- [ ] **Add confidence calibration report** — measure whether confidence scores match actual accuracy (reliability diagram)
- [ ] **Add accuracy regression test to CI** — run `evaluate_predictor` against the fixed test set on every PR; fail if accuracy drops below 0.70
- [ ] **Add feature importance display** — show top k-mer features that drove each prediction in the UI result row

---

## Acceptance Criteria

- [ ] A developer can run one command and get accuracy metrics for the current model against a standard test set
- [ ] The README documents the current accuracy numbers and how to reproduce them
- [ ] The CI pipeline fails if model accuracy regresses below the documented baseline
- [ ] Users can optionally upload labeled sequences and see accuracy metrics in the UI

---

## References

- Evaluation framework: `binary_classifiers/evaluation.py`
- Current metrics: `docs/transformer_classification_guide.md`
- Evaluation tests: `tests/test_binary_classifier_evaluation.py`
- Related issue: #179 (Evo 2 integration — expected to improve accuracy significantly)
